qos/evm/store.go

```diff
@@ -0,0 +1,96 @@
+package evm
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+
+	"github.com/buildwithgrove/path/relayer"
+)
+
+// EndpointStore provides the endpoint selection capability required
+// by the relayer package for handling a service request.
+var _ relayer.EndpointSelector = &EndpointStore{}
+
+// EndpointStoreConfig captures the modifiable settings
+// of the EndpointStore.
+// This will enable `EndpointStore` to be used
+// as part of QoS for other EVM-based blockchains
+// which may have different desired QoS properties.
+// e.g. different blockchains QoS instances could have different
+// tolerance levels for deviation from the current block height.
+type EndpointStoreConfig struct {
+	// TODO_TECHDEBT: apply the sync allowance when validating an endpoint's block height.
+	// SyncAllowance specifies the maximum number of blocks an endpoint
+	// can be behind, compared to the blockchain's estimated block height,
+	// before being filtered out.
+	SyncAllowance uint64
+
+	// ChainID is the ID used by the corresponding blockchain.
+	// It is used to verify responses to service requests with `eth_chainId` method.
+	ChainID string
+}
+
+// EndpointStore maintains QoS data on the set of available endpoints
+// for an EVM-based blockchain service.
+// It performs several tasks, most notable:
+//
+//	1- Endpoint selection based on the quality data available
+//	2- Application of endpoints' observations to update the data on endpoints.
+type EndpointStore struct {
+	Config EndpointStoreConfig
+
+	mutex       sync.RWMutex
```

Comment:
Please update the numae to `endpointsMu` or something like that so it's clear what the mutex is for.